### PR TITLE
fix(workflows): namespace nested descendant step ids in loops/fan-out

### PR DESCRIPTION
**Pull request by Noor-ul-ain001**
## Summary
- `while`/`do-while` loop bodies and `fan-out` templates namespace nested step ids per iteration/item so logs and `state.step_results` entries stay unique — but the namespacing only rewrote the id of the *immediate* child step, not any descendant nested deeper (e.g. a `shell` step inside an `if` inside a `while` body, or inside a `fan-out` template's `if`/`switch` branch). That grandchild kept its bare, unnamespaced id across every iteration/item, so each iteration/item silently overwrote the previous one's entry in `state.step_results` under that same key — only the last iteration's or item's result for that nested step ever survived, and no per-iteration/per-item record of it ever existed.
- This is also a correctness gap beyond bookkeeping: nested/template step ids are deliberately exempted from the workflow's global id-uniqueness validation, on the assumption that runtime namespacing makes any collision safe. Since only the top-level child was actually namespaced, a step nested one level deeper could collide with an unrelated step of the same id elsewhere in the workflow and silently overwrite its result.
- Fix: add `_rename_step_tree_ids`, which recursively rewrites every id in a step's subtree (walking `then`/`else`/`steps`/`default`/`cases.*` — the same nesting keys `overlays/merge.py` walks for step-tree attribution) and returns a `{new_id: original_id}` map. Both the while/do-while loop body and fan-out's `run_item` now use this helper instead of renaming only the top-level id, and alias every renamed descendant's result back to its original id (mirroring the existing single-level aliasing) so sibling steps within the same iteration/item and code reading `steps.<id>.output` after the loop/fan-out still see that iteration's/item's value.

## Follow-up fixes (review rounds)
- Namespace loop iteration 0 too and alias immediately (not after the whole subtree finishes), so the first iteration gets its own recorded entry and a same-iteration sibling sees an earlier sibling's aliased value right away.
- Stop double-renaming loop bodies nested inside an outer loop/fan-out, and stop an unsafe fan-out bare-id alias write from racing across concurrently-running items (each concurrent item now gets a private `context.steps` overlay instead of writing the shared one directly).
- Stop using a `ChainMap` overlay for that per-item isolation: `_resolve_dot_path` (what every `{{ steps.x... }}` expression goes through) only descends through `isinstance(current, dict)`, and `ChainMap` isn't a `dict` subclass, so every such expression inside a concurrent fan-out item silently resolved to `None`. Replaced with a plain `dict` copy.
- Two further edge cases combining previously-separate test scenarios:
  1. *Reserved-id collision + item-local sibling reference*: when a fan-out template id collides with a real, globally-unique step elsewhere in the workflow, the guard that protects that unrelated step's persisted result also suppressed the write needed for a later sibling step *within the same item* to resolve the colliding id locally — the sequential path (which runs directly against the real shared `context.steps`, unlike the concurrent path's private copy) read the *outside* value instead of the item's own. Fixed by always updating `context.steps` for the local read while still skipping the persisted-state write for the collision case; the sequential path now snapshots and restores that entry so the transient item-local write never leaks past the item that made it.
  2. *Nested while/do-while aliasing dropped under concurrency*: a while/do-while body nested inside a fan-out template renames itself dynamically at runtime, so it has no entry in the fan-out's own static id map. The concurrent path's per-item alias publishing was reconstructed from that static map after the item finished, which silently missed the while body's dynamic alias — the namespaced entries (e.g. `fan:item:0:leaf:0`) were still recorded correctly, but the convenience bare-id entry (`leaf`) never was, unlike the sequential path (which publishes immediately during execution instead of reconstructing afterward). Fixed by threading a mutable accumulator through `_execute_steps` that captures every alias write at the point it happens, regardless of nesting depth or whether the rename was static or dynamic.
- **Latest round** — two more issues from Copilot's next review pass:
  1. *Reserved-id restore set too narrow* (landed as a separate 1-line Copilot Autofix commit, then covered here by a regression test): the sequential path's snapshot/restore for a reserved-id collision only covered ids present in the fan-out's own static `id_map`. A nested while/do-while body's id is deliberately absent from that map (it re-namespaces itself dynamically at runtime — see point 2 above), so if such a body's id collided with a reserved id, `_execute_steps`'s collision guard still wrote the item-local value into the live `context.steps[orig_id]`, but nothing restored it afterward — the persisted `state.step_results` entry stayed protected, but the last item's transient value leaked into the live context past the end of the fan-out. Fixed by snapshotting/restoring every id in `context.reserved_step_ids`, not just the ones in `id_map`.
  2. *Nested fan-out doesn't inherit enclosing isolation*: a fan-out template nested inside another fan-out's item decided its own isolation purely from its OWN `max_concurrency`, with no awareness that the ENCLOSING item might itself be one of several running concurrently. A nested fan-out with `max_concurrency <= 1` unconditionally wrote its bare-id alias directly to the real, persisted `state.step_results` as each inner item completed — so two outer items running that same nested fan-out concurrently could race on that key, the exact race the outer fan-out's own isolation exists to prevent. It also never fed its aliases into the enclosing item's own accumulator, so an outer sibling step could never see them. Fixed by threading the enclosing `_execute_steps` call's `alias_local_only`/`alias_records` through `_run_fan_out` as `parent_local_only`/`parent_alias_records`: every item of a nested fan-out (sequential or concurrent) now isolates when the enclosing item does, and its aliases flow into the enclosing accumulator instead of the persisted store.

## Test plan
- [x] `test_while_loop_namespaces_nested_descendant_steps`, `test_fan_out_namespaces_nested_descendant_steps` (original fix)
- [x] `test_fan_out_concurrent_sibling_step_isolated_per_item`, `test_fan_out_concurrent_sibling_step_resolves_via_expression`, `test_fan_out_concurrent_alias_never_clobbers_unrelated_step`, `test_fan_out_sequential_alias_never_clobbers_unrelated_step` (prior review rounds)
- [x] **New this round**: `test_fan_out_reserved_id_collision_still_resolves_item_local_sibling`, parametrized over `max_concurrency` 1 and 2 — a template's second step reads the first's output via `{{ steps.first.output... }}` where `first` collides with an outside reserved step; asserts the item-local value resolves under both paths, and the outside step's entry survives untouched
- [x] **New this round**: `test_while_loop_nested_in_fan_out_aliases_to_true_original_id` is now parametrized over `max_concurrency` 1 and 2 (previously sequential-only)
- [x] Verified both new/updated tests fail without this round's fix (test-the-test, `git stash` on `engine.py` only): `test_fan_out_reserved_id_collision_still_resolves_item_local_sibling[1]` failed with the outside value (`'unrelated' == 'a'`) leaking into the item-local read; `test_while_loop_nested_in_fan_out_aliases_to_true_original_id[2]` failed with `KeyError: 'leaf'` — both reproducing exactly the two symptoms from review, then passed once restored
- [x] Ran the full `tests/test_workflows.py` suite (prior round): 935 passed, 20 pre-existing Windows symlink-elevation failures (need admin rights, unrelated to this change), 7 skipped — no regressions
- [x] **New this round**: `test_nested_fan_out_inside_concurrent_fan_out_publishes_by_item_order` — outer fan-out runs 2 items concurrently, each item's template is itself a fan-out (`max_concurrency: 1`) ending in a bare-id `leaf` step. Forces outer item "x" (item order 0) to finish strictly *after* outer item "y" (item order 1) via a `threading.Event`, then asserts `state.step_results["leaf"]` still resolves to "y" (the item-order convention) rather than "x" (which a completion-order race would produce)
- [x] **New this round**: `test_fan_out_sequential_nested_while_reserved_collision_restores_live_context` — a while/do-while body's dynamically-namespaced id collides with a reserved outside id; asserts `context.steps["leaf"]` is restored to the outside value after the fan-out finishes, not left holding the last item's transient value
- [x] Verified both new tests fail without their respective fix (test-the-test): the nested-fan-out test asserted `'x' == 'y'` and failed with `x` when `_run_fan_out` didn't accept `parent_local_only`; the while-collision test failed with the last item's value (`'c'`) leaking into `context.steps["leaf"]` when the restore loop was reverted to `set(id_map.values()) & context.reserved_step_ids` (the pre-Autofix form) — both passed once restored
- [x] Ran the full `tests/test_workflows.py` suite again this round: 937 passed, 20 pre-existing Windows symlink-elevation failures (unrelated), 7 skipped — no regressions
- [x] `ruff check .` clean

## AI Disclosure

- [x] I **did** use AI assistance (describe below)

This PR, across all rounds including this one, was authored with [Claude Code](https://claude.com/claude-code) (Anthropic CLI), model Claude Sonnet 5, used in its interactive, agentic CLI mode with full local tool access (file read/edit, running `pytest`, `ruff`, `git`). For this round specifically: the AI agent traced both new review comments back to their root cause in `_execute_steps`/`_run_fan_out` (confirming the reserved-id restore gap was already fixed by a separate Copilot Autofix commit already on this branch, and writing the isolation-propagation fix and both regression tests above), ran the test-the-test verification and full suite described above, and wrote this description, under my direction. I reviewed the diff and test output myself before pushing and understand the change being made.

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>

### CODE REVIEW

**Comment by mnriem** [comment]
Please address Copilot feedback

**Comment by mnriem** [comment]
Strong direction — this is a real correctness bug (silent per-iteration result overwrite + the uniqueness-exemption collision risk), disclosed and tested. Two review findings to resolve before merge: the **first loop iteration still isn't namespaced** (it goes through the generic `result.next_steps` path before `_rename_step_tree_ids` runs), and the **post-subtree aliasing breaks sibling references and concurrent fan-out isolation** — a later child can execute before an earlier renamed child is aliased back. Please address those and re-request review.

Separately: you currently have 7 open PRs. Per CONTRIBUTING, that's past the point where further submissions may be deprioritized in the queue. Several are small workflow/config validation fixes (#4319, #4323, #4324, #4325) — could you **consolidate the related ones** so review can focus? This PR can stay standalone; it's the smaller validation fixes I'd group.

File: src/specify_cli/workflows/engine.py [hunk 1]
```diff
@@ -1457,18 +1564,65 @@ def item_id(idx: int) -> str:
...
+            # shared steps dict that other concurrently-running items also
+            # read from — the actual race Copilot flagged: every worker
+            # writing the same bare-id key could otherwise expose another
+            # item's value to a sibling read. The caller applies exactly
+            # one item's aliases to shared state — deterministically the
+            # last item in item order — once every item has finished.
+            original_steps = item_ctx.steps
+            local_overlay: dict[str, dict[str, Any]] = {}
+            if local_only:
+                item_ctx.steps = ChainMap(local_overlay, original_steps)
```
[the remark this answers is not in the document]
Comment (Noor-ul-ain001) [reply src/specify_cli/workflows/engine.py]: Confirmed and fixed. Verified directly: `isinstance(ChainMap(...), dict)` is `False`, and `_build_namespace`'s `ns["steps"] = context.steps or {}` put that ChainMap straight into the expression namespace, so `_resolve_dot_path`'s `isinstance(current, dict)` check failed on the very first `steps.<id>` hop -- every such expression inside a concurrent fan-out item silently resolved to `None`. Replaced the ChainMap with a plain `dict(original_steps)` snapshot taken once per item: still isolates this item's own writes from concurrently-running siblings (a fresh copy, never shared), but is a real dict so every expression path works. Added a test that goes through `evaluate_expression("{{ steps.first.output.marker }}", context)` (the real templating path) rather than `context.steps.get()`, and confirmed via test-the-test that it fails against the ChainMap version (`None == 0`) and passes now (a8188b60).

File: src/specify_cli/workflows/engine.py [hunk 2]
```diff
@@ -1231,6 +1321,13 @@ def _execute_steps(
                 "error": result.error,
             }
             self._record_result(context, state, step_id, step_data)
+            if alias_map is not None:
+                orig_id = alias_map.get(step_id)
+                if orig_id is not None:
+                    if alias_local_only:
+                        context.steps[orig_id] = step_data
+                    else:
+                        self._record_result(context, state, orig_id, step_data)
```
[the remark this answers is not in the document]
Comment (Noor-ul-ain001) [reply src/specify_cli/workflows/engine.py:1407]: Fixed both this (sequential) and the concurrent post-join write at line 1751/now further down. Added `_collect_reserved_step_ids`, which walks the workflow's regular control-flow nesting keys (then/else/steps/default/cases) -- the same shape `_validate_steps` walks into its global `seen_ids` -- but deliberately skips a fan-out's own `step` template, mirroring the fresh-`set()` exemption at engine.py:569-580. That gives a static, run-wide set of every id declared OUTSIDE a fan-out template. Both alias-write sites (this one and the concurrent post-join loop) now skip writing the bare-id alias when `orig_id` is a member of that set -- the namespaced entry is written unconditionally either way, so nothing is lost, only the unsafe convenience alias is suppressed. Kept it scoped to fan-out specifically (a new `alias_may_collide` flag, set only for fan-out template execution) rather than checking every alias write, since a while/do-while loop body's ids ARE globally unique by validation and gating those too would have broken the loop-aliasing this PR adds. Added tests for both the sequential and concurrent paths, each confirmed via test-the-test to fail pre-fix (the unrelated step's result gets clobbered with the fan-out's last item's data) and pass now (7365250c).

**Comment by Noor-ul-ain001** [comment]
Addressed everything in this review — both the two posted inline comments and the two suppressed ones summarized in the review body.

**1. Double-renaming when a while/do-while loop is nested inside an outer loop iteration or fan-out item** (the suppressed comment quoting `engine.py:1457`): confirmed exactly as described — a nested while step's own `steps` body was being eagerly renamed by the outer `_rename_step_tree_ids` pass (since `steps` is a walked nesting key), so the loop's own per-iteration rename then treated that already-namespaced id as "original" and aliased back to a synthetic id like `fan:leaf:0` instead of the real `leaf`. `state.step_results["leaf"]` was never populated at all. Fixed by having `_rename_step_tree_ids` still rename a nested while/do-while step's own id, but leave its `steps` body untouched — the loop renames it exactly once, against the real original ids, when it actually runs.

**2 & 3. Fan-out bare-id alias clobbering an unrelated step** (both the suppressed concurrent-path comment quoting `engine.py:1752`, and the posted sequential-path comment): fan-out template ids are exempt from the workflow's global id-uniqueness check, so a template's bare id can coincide with a real step elsewhere. Added `_collect_reserved_step_ids` (mirrors `_validate_steps`'s global `seen_ids`, minus the fan-out-template exemption) and gated both the sequential path's immediate alias write and the concurrent path's deferred post-join write on membership in that set — scoped via a new `alias_may_collide` flag so ordinary loop-body aliasing (always globally unique) is untouched.

**4. `ChainMap` breaking `steps.*` expressions in concurrent fan-out** (posted comment): confirmed — `isinstance(ChainMap(...), dict)` is `False`, and `_resolve_dot_path` only descends through actual dicts, so every `{{ steps.x.output... }}` expression inside a concurrent fan-out item silently resolved to `None`. Replaced the `ChainMap` overlay with a plain `dict` snapshot per item, which keeps the same isolation property while being a real dict.

Each fix has a dedicated regression test, and each was confirmed via test-the-test (temporarily reverting just that fix) to fail against the prior code and pass now. Full `test_workflows.py` run clean except the pre-existing Windows-symlink-elevation failures unrelated to this change (20 failures, same set as baseline on `main`). Pushed as 7365250c and a8188b60.

**Comment by mnriem** [comment]
The earlier review fixes are present, but two combined cases still need correction.

First, combine the reserved-ID collision case with a sibling reference inside the fan-out item. With an outside `first` result and an item-local `first`, sequential execution reads the outside value while concurrent execution correctly reads the item’s value. Protecting shared state must not suppress item-local lookup.


Second, run the nested-while test with both one and two workers. Sequential execution publishes the latest bare `leaf` result; concurrent execution leaves it absent, despite preserving the namespaced records. Please carry those nested aliases through deterministic publication.


Please cover both cases across worker counts and complete the plain AI disclosure with tool, mode/settings, and extent of assistance; the model attribution is already present.


*Drafted for @mnriem with assistance from GitHub Copilot (model: GPT-6 Astra; interactive comment drafting).*

**Comment by Noor-ul-ain001** [comment]
Thanks for combining those two cases — both were real, and each was masked by the other's path working correctly:

1. **Reserved-id collision + item-local sibling reference**: the guard that protects an outside colliding step's persisted result was also suppressing the `context.steps` write a same-item sibling needs for local resolution — only on the sequential path, since it runs directly against the real shared `context.steps` (the concurrent path already has a private per-item copy, which is why it was unaffected). Fixed by always writing `context.steps` for the local read while still skipping the persisted-state write for the collision case, with a snapshot/restore in `run_item` so that transient write never leaks past the item that made it.
2. **Nested while/do-while aliasing dropped under concurrency**: a nested while body renames itself dynamically at runtime, so the concurrent path's post-hoc reconstruction of alias records from the fan-out's *static* id map missed it — the namespaced entries were fine, only the bare-id convenience alias (`leaf`) was silently dropped. Fixed by threading a mutable accumulator through `_execute_steps` that records every alias at write time, at any nesting depth, instead of reconstructing afterward.

Added `test_fan_out_reserved_id_collision_still_resolves_item_local_sibling` and parametrized the existing `test_while_loop_nested_in_fan_out_aliases_to_true_original_id` over `max_concurrency` 1 and 2, as requested. Verified both fail on the pre-fix code with the exact symptoms described (outside value leaking into the item-local read; `KeyError: 'leaf'`), then pass with the fix. Full suite: 935 passed, same 20 pre-existing Windows symlink failures, 7 skipped.

Also added a plain AI Disclosure section to the PR body (tool/model, mode, extent of assistance) covering this round and the prior ones.

**Comment by Noor-ul-ain001** [comment]
Addressed both findings from the latest Copilot review round.

**1. Reserved-id restore set too narrow** — this was already fixed by the Copilot Autofix commit that landed directly on this branch (`7370cb2`, "Refactor loop to iterate directly over `reserved_step_ids`"): the sequential path's snapshot/restore now iterates every id in `context.reserved_step_ids` instead of only the ones present in the fan-out's static `id_map`, so a nested while/do-while body's dynamically-namespaced id colliding with a reserved id is restored in the live context too. I added a regression test for it (`test_fan_out_sequential_nested_while_reserved_collision_restores_live_context`) and verified it fails when reverted to the pre-Autofix form (`context.steps["leaf"]` was left holding the last item's transient value instead of being restored).

**2. Nested fan-out bypasses outer concurrency isolation** — confirmed exactly as described: `_run_fan_out` decided isolation purely from its own `max_concurrency`, with no way to know it was itself running inside an already-isolated enclosing item. A nested fan-out with `max_concurrency <= 1` wrote its bare-id alias straight to the persisted, truly-shared `state.step_results` as each inner item completed, so two outer items concurrently running that same nested fan-out could race on that key — and the write never reached the enclosing item's own `alias_records` accumulator either.

Fixed by threading the enclosing `_execute_steps` call's `alias_local_only`/`alias_records` into `_run_fan_out` as new `parent_local_only`/`parent_alias_records` keyword args. Every item of a nested fan-out — sequential or concurrent — now isolates whenever the enclosing item does, and its aliases flow into the enclosing accumulator instead of the persisted store, mirroring how a nested while/do-while body's dynamic alias already reaches that accumulator.

Added `test_nested_fan_out_inside_concurrent_fan_out_publishes_by_item_order`: outer fan-out runs 2 items concurrently, each item's template is itself a fan-out (`max_concurrency: 1`, i.e. its own decision would be "no isolation needed") ending in a bare-id `leaf` step. A `threading.Event` forces outer item "x" (item order 0) to finish strictly *after* outer item "y" (item order 1) despite both starting concurrently. Verified via test-the-test: pre-fix this asserts `'x' == 'y'` and gets `'x'` (whichever outer item's write happens to land last in wall-clock time wins the race); post-fix it deterministically resolves to "y", matching the item-order convention every other fan-out alias in this module follows.

Full `tests/test_workflows.py` suite: 937 passed, 20 pre-existing Windows symlink-elevation failures (unrelated), 7 skipped. `ruff check .` clean.

**Comment by Noor-ul-ain001** [comment]
Addressed the open findings from the last two Copilot reviews in 2782bdfc.

1. **Syntax error from the Autofix commit (1088237f).** Its lines landed at column 0 inside the sequential fan-out loop, so `engine.py` no longer parsed. I re-indented them and kept what the Autofix meant to do (the "publish nested fan-out aliases to the private context view" finding): a sequential fan-out nested inside an isolated outer item now writes its aliases into that item's private `context.steps`, as well as its accumulator. A later sibling in the same outer item can then resolve `steps.<inner-id>`.
2. **Unlocked write to persisted state on resume.** On resume, `context.steps` is `state.step_results`. `run_item`'s namespaced mirror now runs only when the two are distinct dicts. `_record_result` has already written those keys under the run lock, so the mirror was a redundant unlocked write that could race another worker's `save()`.
3. **Aliases from earlier items dropped (the "previously missed" finding).** The concurrent path published only the last collected item's alias map. It now folds every collected item's map, up to the halting item, in item order. Each bare id then resolves to the latest item that actually ran it, the same result as the sequential path.

New tests:
- `test_sequential_fan_out_nested_in_concurrent_item_visible_to_later_sibling`
- `test_fan_out_alias_kept_from_earlier_item_not_run_by_last` (parametrized over `max_concurrency` 1 and 2)
- `test_concurrent_fan_out_on_resume_never_writes_step_results_unlocked`. This one swaps in a lock that records its owning thread, so it detects the unlocked write deterministically instead of relying on the race's timing.

I reverted each fix on its own and confirmed its test failed: `None == 'p'`, `['fan:leaf:0', 'fan:leaf:1'] == []`, and `KeyError: 'leaf'` (concurrent case only). Full `tests/test_workflows.py`: 941 passed, 7 skipped, and 20 failures. All 20 are the same Windows symlink-privilege failures (`WinError 1314`) that fail on `main`. `ruff check .` is clean.

**Comment by mnriem** [comment]
Please address Copilot feedback and fix test & lint errors

**Comment by Noor-ul-ain001** [comment]
Addressed in 54a03a2f (plus merge of `upstream/main`, 6f4910fe):

- **Test failures:** the 8 failing tests all raised `ModuleNotFoundError: specify_cli.workflows.steps`. Upstream renamed that package to `specify_cli.workflows.step`, and CI tests the merge result. I merged `upstream/main` and updated this PR's imports.
- **Copilot "Nested fan-out reset discards enclosing context.item":** removed the `context.item = None` after the `_run_fan_out` call in `_execute_steps`. `_run_fan_out` already restores the caller's item on the sequential path and never mutates it on the concurrent path. Added `test_nested_fan_out_keeps_enclosing_item_for_later_sibling`, which covers outer and inner concurrency of 1 and 2. All 4 cases fail without the fix (`assert None == 'x'`) and pass with it.
- **Lint:** `ruff check src tests` is clean.

Local run of `tests/test_workflows.py` and `tests/workflows/`: 904 passed, 5 skipped. The 11 failures are the symlink tests, which need elevation on Windows and are unrelated.

File: src/specify_cli/workflows/engine.py [hunk 3]
```diff
@@ -925,6 +925,135 @@ def append_log(self, entry: dict[str, Any]) -> None:
...
+                    else:
+                        renamed_cases.append(child)
+                new_cases[case_key] = renamed_cases
+            else:
+                new_cases[case_key] = case_steps
+        new_step["cases"] = new_cases
+    return new_step, id_map
+
+
+def _collect_reserved_step_ids(steps: list[dict[str, Any]]) -> frozenset[str]:
```
[the remark this answers is not in the document]
Comment (Noor-ul-ain001) [reply src/specify_cli/workflows/engine.py]: Added `test_fan_out_reserved_id_collision_end_to_end` in 1422af02, parametrized over `max_concurrency` 1 and 2. It runs a real workflow through `engine.execute()`: an outside `leaf` step, then a fan-out whose template also uses id `leaf`, then an `after` step reading `steps.leaf.output.stdout`. It asserts the outside result survives in `state.step_results` and in the live context. Confirmed both variants fail (`'c
' == 'outside
'`) when the `reserved_step_ids` wiring in `execute()` is removed.

Posted on behalf of @Noor-ul-ain001 by Claude Code (model: claude-opus-5-5).
